docs/general/builtins/table-functions.rst

```diff
@@ -140,6 +140,94 @@ The return value always matches the ``start`` / ``stop`` types.
     +---------------+-----------------------------------+
     SELECT 3 rows in set (... sec)
 
+.. _table-functions-generate-subscripts:
+
+``pg_catalog.generate_subscripts(array, dim, [reverse])``
+=========================================================
+
+Generate the subscripts for the specified dimension ``dim`` of the given
+``array``. Zero rows are returned for arrays that do not have the requested
+dimension, or for NULL arrays (but valid subscripts are returned for NULL
+array elements).
+
+If ``reverse`` is ``true`` the subscripts will be returned in reverse order.
+
+This example takes a one dimensional array of four elements, where elements
+at positions 1 and 3 are NULL:
+
+::
+
+    cr> SELECT generate_subscripts([NULL, 1, NULL, 2], 1) AS s;
+    +---+
+    | s |
+    +---+
+    | 1 |
+    | 2 |
+    | 3 |
+    | 4 |
+    +---+
+    SELECT 4 rows in set (... sec)
+
+This example returns the reversed list of subscripts for the same array:
+
+::
+
+    cr> SELECT generate_subscripts([NULL, 1, NULL, 2], 1, true) AS s;
+    +---+
+    | s |
+    +---+
+    | 4 |
+    | 3 |
+    | 2 |
+    | 1 |
+    +---+
+    SELECT 4 rows in set (... sec)
+
+The following examples work on an array of three dimensions. Each of the elements
+within a given level must be either NULL, or an array of the same size as the other
+arrays within the same level.
```

Comment:
I wonder if we should actually be a bit more lenient about this and take the longest array at the given dimension?

In PostgreSQL this is a limitation of the array type itself:

```
=# select array[ array[1, 2],  array[3, 4, 5] ];
ERROR:  2202E: multidimensional arrays must have array expressions with matching dimensions
```

@kovrus what do you think?

Reply:
The longest array would result  in `index out of bounds` errors.
The shortest array would result in missing elements.